pkg/placement/placement-decision.go

```diff
@@ -0,0 +1,144 @@
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package placement
+
+import (
+	"context"
+	"fmt"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+
+	"github.com/kubestellar/kubestellar/api/edge/v1alpha1"
+	"github.com/kubestellar/kubestellar/pkg/util"
+)
+
+// syncPlacementDecision syncs a placement-decision object with what is resolved by the placement-resolver.
+func (c *Controller) syncPlacementDecision(key util.Key) error {
+	obj, err := c.getObjectFromKey(key)
+	if err != nil {
+		// The resource doesn't exist, needs to be created
+		obj = util.EmptyUnstructuredObjectFromKey(key)
+	}
+
+	unstructuredObj, ok := obj.(*unstructured.Unstructured)
+	if !ok {
+		return fmt.Errorf("failed to convert runtime.Object to unstructured.Unstructured")
+	}
+
+	placementDecision, err := unstructuredObjectToPlacementDecision(unstructuredObj)
+	if err != nil {
+		return fmt.Errorf("failed to convert from Unstructured to PlacementDecision: %w", err)
+	}
+
+	// placement decision name matches that of the placement 1:1, therefore its NamespacedName is the same.
+	placementIdentifier := placementDecision.GetName()
+
+	// generate placement decision spec from resolver
+	generatedPlacementDecisionSpec, err := c.placementResolver.GeneratePlacementDecision(placementIdentifier)
+	if err != nil {
+		return fmt.Errorf("failed to get placement decision spec: %w", err)
+	}
+
+	// calculate if the resolved decision is different from the current one
+	if !c.placementResolver.ComparePlacementDecision(placementIdentifier, &placementDecision.Spec) {
+		// update the placement decision object in the cluster by updating spec
+		if err = c.updateOrCreatePlacementDecision(placementDecision, generatedPlacementDecisionSpec); err != nil {
+			return fmt.Errorf("failed to update or create placement decision: %w", err)
+		}
+	}
+
+	return nil
+}
+
+// updateOrCreatePlacementDecision updates or creates a placement-decision object in the cluster.
+// If the object already exists, it is updated. Otherwise, it is created.
+func (c *Controller) updateOrCreatePlacementDecision(pd *v1alpha1.PlacementDecision,
+	placementDecisionSpec *v1alpha1.PlacementDecisionSpec) error {
+	// use the passed placement decision and set its spec
+	pd.Spec = *placementDecisionSpec
+
+	// set owner reference
+	//	get owner placement object
+	placement, err := c.getPlacementByName(pd.GetName())
+	if err != nil {
+		return fmt.Errorf("failed to get placement %s: %w", pd.GetName(), err)
+	}
```

Comment:
Since there is already a `PlacementDecisionSpec`, this implies that the corresponding `Placement` has already been read. Why read it again instead of remembering the information needed for the OwnerReference?

Reply:
I thought you were previously referring to why I'm making an API call to read the placement instead of using the lister's cache, since the placement has been read for sure at a previous step.

Two problems:
* If the spec was just made in-memory (for initial creation), it wouldnt have an owner-reference in the current code, which can be solved by adding one but then we are left with problem 2. 
* Dont we need to verify that the ownerReference is correct and was not modified by anyone?

If you think it's sufficient to set the ownerReference when creating the object (it would still require reading the placement though, but once), I can do that.

Reply:
Let's back up and talk about PlacementResolver. The methods NoteObject, RemoveObject, and SetDestinations are only called with a Placement in hand. I am surprised by the comment "If a key does not exist in the cache, it is created" on the GeneratePlacementDecision method. What goes wrong if instead of creating empty content for an absent key, this method returns a `nil` value for the `*v1alpha1.PlacementDecisionSpec`? If a Placement has not been seen, or has been seen _and_ deleted since then, it seems reasonable that this controller should not be trying to write a PlacementDecision value.

Reply:
Also: why enqueue a distinct item to make this controller work on a PlacementDecision? Wouldn't it be more natural --- and concurrency-safe --- to do the work on a Placement's decision while syncing a reference to that Placement object?

Reply:
For RemoveObject I've updated it to not create an empty resolution if none is associated with the placementKey. 

For GeneratePlacementDecision, I admit I didn't really try to simulate all scenarios to figure out what is safe and what isn't, and went for the safe option of: if someone is requesting a generated spec for a resolution that doesn't exist, then the spec should be empty.
Though what I just said can be enforced by the caller of Generate, so I'll implement your suggestion.

Reply:
As I understand it, the idea is that syncing a reference to a PlacementDecision is supposed to look in the PlacementResolver to pull out what has been stashed there while syncing other objects and make the external representation in the apiserver(s) match what the PlacementResolver said. When the PlacementResolver has nothing for a given Placement name, this means that the apiserver(s) should hold no corresponding PlacementDecision object.

Think carefully about concurrency and the PlacementResolver. Is it possible that goroutine A calls DeleteResolution for a given placement and then later goroutine B calls NoteObject while working with older information? Yes, it is. That's a bug.

Reply:
Solved in https://github.com/kubestellar/kubestellar/pull/1681/commits/8af23c51daeb8c1267271dbf5e0badee87bd6a76

The solution is to stop allowing these functions to create resolutions if missing, and to sync up empty resolutions for existing placements on controller load. The alternative was to implement some form of resolution validation that allows for syncing.

Reply:
Done